benchmarks/CMakeLists.txt

```diff
@@ -0,0 +1,74 @@
+# Algebra plugins library, part of the ACTS project (R&D line)
+#
+# (c) 2023 CERN for the benefit of the ACTS project
+#
+# Mozilla Public License Version 2.0
+
+# Set the default C++ compiler flags.
+include( algebra-plugins-compiler-options-cpp )
+
+# Set up an INTERFACE library for the common header(s).
+add_library( algebra_bench_common INTERFACE )
+target_include_directories( algebra_bench_common INTERFACE
+   "${CMAKE_CURRENT_SOURCE_DIR}/common/include" )
+if( "${CMAKE_CXX_COMPILER_ID}" MATCHES "MSVC" )
+   target_compile_definitions( algebra_bench_common INTERFACE
+      -D_USE_MATH_DEFINES )
+endif()
+add_library( algebra::bench_common ALIAS algebra_bench_common )
+
+# Set up all of the (available) "host" benchmarks.
+add_library( algebra_bench_array INTERFACE )
+target_include_directories( algebra_bench_array INTERFACE 
+   "$<BUILD_INTERFACE:${CMAKE_CURRENT_SOURCE_DIR}/array/include>" )
+target_link_libraries(algebra_bench_array INTERFACE algebra::array_cmath 
+                                                    algebra::common_storage)
+add_library( algebra::bench_array ALIAS algebra_bench_array )
+
+algebra_add_benchmark( array_getter
+   "array/array_getter.cpp"
+   LINK_LIBRARIES benchmark::benchmark algebra::bench_common
+                   algebra::bench_array algebra::array_cmath )
+algebra_add_benchmark( array_vector
+   "array/array_vector.cpp"
+   LINK_LIBRARIES benchmark::benchmark algebra::bench_common
+                   algebra::bench_array algebra::array_cmath )
+
+if( ALGEBRA_PLUGINS_INCLUDE_EIGEN )
+   add_library( algebra_bench_eigen INTERFACE )
+   target_include_directories( algebra_bench_eigen INTERFACE 
+      "$<BUILD_INTERFACE:${CMAKE_CURRENT_SOURCE_DIR}/eigen/include>" )
+   target_link_libraries(algebra_bench_eigen INTERFACE algebra::eigen_eigen
+                                                    algebra::common_storage)
+   add_library( algebra::bench_eigen ALIAS algebra_bench_eigen )
+
+   algebra_add_benchmark( eigen_getter
+      "eigen/eigen_getter.cpp"
+      LINK_LIBRARIES benchmark::benchmark algebra::bench_common
+                     algebra::bench_eigen algebra::eigen_eigen )
+   algebra_add_benchmark( eigen_vector
+      "eigen/eigen_vector.cpp"
+      LINK_LIBRARIES benchmark::benchmark algebra::bench_common
+                     algebra::bench_eigen algebra::eigen_eigen )
+endif()
+
+if( ALGEBRA_PLUGINS_INCLUDE_VC )
+   if( NOT "${CMAKE_CXX_COMPILER_ID}" MATCHES "AppleClang" )
```

Comment:
I know that I opened a thread about this somewhere already, but couldn't find it. So here we go with a new one.

When I saw that Fastor was having problems on my M1 Mac, I was absolutely convinced that that would be the reason behind the CI's troubles with the benchmarks for the SoA code. But as it turns out, no. 😕

```
[zsh][Thranduil]:algebra > ./algebra-plugins/out/build/vc/bin/algebra_benchmark_vc_soa_getter
Algebra-Plugins 'getter' benchmark (Vc SoA)
-------------------------------------------

(single)
 -> running:	 160000 samples
 -> warmup: 	 16000 samples

(double)
 -> running:	 160000 samples
 -> warmup: 	 16000 samples

Unable to determine clock rate from sysctl: hw.cpufrequency: No such file or directory
2023-10-14T07:32:48+02:00
Running ./algebra-plugins/out/build/vc/bin/algebra_benchmark_vc_soa_getter
Run on (8 X 24.1209 MHz CPU s)
CPU Caches:
  L1 Data 64 KiB (x8)
  L1 Instruction 128 KiB (x8)
  L2 Unified 4096 KiB (x2)
Load Average: 1.61, 1.61, 2.29
-------------------------------------------------------------------------------------------
Benchmark                                                 Time             CPU   Iterations
-------------------------------------------------------------------------------------------
vector_GETTER_NAME_single/process_time/threads:8     216412 ns      1500595 ns          360
vector_GETTER_NAME_double/process_time/threads:8     717059 ns      5376955 ns          136
vector_GETTER_NAME_single/process_time/threads:8     218412 ns      1630665 ns          472
vector_GETTER_NAME_double/process_time/threads:8     807922 ns      6015875 ns          120
vector_GETTER_NAME_single/process_time/threads:8      19488 ns       146199 ns         5288
vector_GETTER_NAME_double/process_time/threads:8      19847 ns       147177 ns         5936
vector_GETTER_NAME_single/process_time/threads:8      20878 ns       148881 ns         4960
vector_GETTER_NAME_double/process_time/threads:8      21053 ns       152908 ns         5160
vector_GETTER_NAME_single/process_time/threads:8     275938 ns      2077637 ns          360
vector_GETTER_NAME_double/process_time/threads:8     394006 ns      2770471 ns          288
[zsh][Thranduil]:algebra > ./algebra-plugins/out/build/vc/bin/algebra_benchmark_vc_soa_vector
Algebra-Plugins 'vector' benchmark (Vc SoA)
-------------------------------------------

(single)
 -> running:	 160000 samples
 -> warmup: 	 16000 samples

(double)
 -> running:	 160000 samples
 -> warmup: 	 16000 samples

Unable to determine clock rate from sysctl: hw.cpufrequency: No such file or directory
2023-10-14T07:32:52+02:00
Running ./algebra-plugins/out/build/vc/bin/algebra_benchmark_vc_soa_vector
Run on (8 X 24.1208 MHz CPU s)
CPU Caches:
  L1 Data 64 KiB (x8)
  L1 Instruction 128 KiB (x8)
  L2 Unified 4096 KiB (x2)
Load Average: 1.56, 1.60, 2.28
-----------------------------------------------------------------------------------------
Benchmark                                               Time             CPU   Iterations
-----------------------------------------------------------------------------------------
vector_add_single/process_time/threads:8            29176 ns       215771 ns         2440
vector_add_double/process_time/threads:8            33332 ns       249757 ns         2912
vector_sub_single/process_time/threads:8            28771 ns       210642 ns         3688
vector_sub_double/process_time/threads:8            33545 ns       246659 ns         3304
vector_dot_single/process_time/threads:8            27937 ns       200166 ns         3808
vector_dot_double/process_time/threads:8            33460 ns       247557 ns         2920
vector_cross_single/process_time/threads:8          32187 ns       241733 ns         3064
vector_cross_double/process_time/threads:8          36558 ns       275813 ns         2728
vector_normalize_single/process_time/threads:8      38788 ns       293758 ns         2664
vector_normalize_double/process_time/threads:8      40530 ns       298386 ns         2416
[zsh][Thranduil]:algebra >
```

My M1 chip seems to be just fine with these instructions. Much because Vc doesn't actually do anything on an ARM chip at the moment.

```
-- Building Vc as part of the Algebra Plugins project
-- Detected Compiler: Clang 15.0.0
CMake Warning at /Users/krasznaa/ATLAS/projects/algebra/build/_deps/vc-src/CMakeLists.txt:34 (message):
  No optimized implementation of the Vc types available for arm64
```

I'll revive my old MacBook for this, just to see if I can even reproduce the issues that you've been seeing in the CI... 🤔

Reply:
It seemed to fail only in release mode, if I remember correctly, so only on specific instructions

Reply:
Could you try re-enabling it in the CI, just so that I could see how it fails? I can't reproduce it on my 2014 MacBook Pro either.

The issue is either gone through some recent modification, or it's very specific to the CPU used in the GitHub runners. :confused:

Reply:
Is the code being compiled for a native target by default? ARM chips have a lot more variety in what features they do and don't support which makes it extra important to compile for the right target.

Reply:
Can you try printing `Vc_COMPILE_FLAGS`, `Vc_ARCHITECTURE_FLAGS` and `cat /proc/cpuinfo |grep flags`?

I know Vc does not support ARM, and I suspect that it is probably not setting any architecture flag at all, or it is setting some incompatible flag.

You could also try adding a CXXFLAGS="-march=native"` environment flag and let the compiler figure it out.

Reply:
Or you could try `CXXFLAGS="-march=armv8-a"` and see if that works.

Reply:
I don't know if this is helps, but during configuration of Vc the XCode CI prints:
```
-- Detected CPU: ivy-bridge
-- Performing Test check_cxx_compiler_flag__march_ivybridge
-- Performing Test check_cxx_compiler_flag__march_ivybridge - Success
```

Reply:
Oh I thought this was about ARM. :laughing: Never mind.

Reply:
Unfortunately I'm running out of ideas of how to debug this further. 😦

The CI nodes used macOS 12.7 and Xcode 14. On my old MacBook the latest software that I can install is macOS 11.7 and Xcode 13. 😕 And as the Vc configuration reports, my laptop has a Haswell CPU. While the CI runs on Ivy Bridge Xeon CPUs as it seems.

As far as I can see, the CI uses the same optimization settings for the build

```
CompileC /Users/runner/work/algebra-plugins/algebra-plugins/build/build/algebra_test_vc_soa.build/Release/Objects-normal/x86_64/vc_soa.o /Users/runner/work/algebra-plugins/algebra-plugins/tests/vc_soa/vc_soa.cpp normal x86_64 c++ com.apple.compilers.llvm.clang.1_0.compiler (in target 'algebra_test_vc_soa' from project 'algebra-plugins')
    cd /Users/runner/work/algebra-plugins/algebra-plugins
    /Applications/Xcode_14.2.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin/clang -x c++ -target x86_64-apple-macos12.7 -fmessage-length\=0 -fdiagnostics-show-note-include-stack -fmacro-backtrace-limit\=0 -Wno-trigraphs -fpascal-strings -O3 -Wno-missing-field-initializers -Wno-missing-prototypes -Wno-return-type -Wno-non-virtual-dtor -Wno-overloaded-virtual -Wno-exit-time-destructors -Wno-missing-braces -Wparentheses -Wswitch -Wno-unused-function -Wno-unused-label -Wno-unused-parameter -Wno-unused-variable -Wunused-value -Wno-empty-body -Wno-uninitialized -Wno-unknown-pragmas -Wno-shadow -Wno-four-char-constants -Wno-conversion -Wno-constant-conversion -Wno-int-conversion -Wno-bool-conversion -Wno-enum-conversion -Wno-float-conversion -Wno-non-literal-null-conversion -Wno-objc-literal-conversion -Wshorten-64-to-32 -Wno-newline-eof -Wno-c++11-extensions -Wno-implicit-fallthrough -DCMAKE_INTDIR\=\"Release\" -isysroot /Applications/Xcode_14.2.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX13.1.sdk -fasm-blocks -fstrict-aliasing -Wdeprecated-declarations -Winvalid-offsetof -Wno-sign-conversion -Wno-infinite-recursion -Wno-move -Wno-comma -Wno-block-capture-autoreleasing -Wno-strict-prototypes -Wno-range-loop-analysis -Wno-semicolon-before-method-body -I/Users/runner/work/algebra-plugins/algebra-plugins/build/bin/Release/include -I/Users/runner/work/algebra-plugins/algebra-plugins/storage/vc_soa/include -I/Users/runner/work/algebra-plugins/algebra-plugins/common/include -I/Users/runner/work/algebra-plugins/algebra-plugins/storage/common/include -I/Users/runner/work/algebra-plugins/algebra-plugins/frontend/vc_soa/include -I/Users/runner/work/algebra-plugins/algebra-plugins/math/cmath/include -I/Users/runner/work/algebra-plugins/algebra-plugins/math/common/include -I/Users/runner/work/algebra-plugins/algebra-plugins/math/vc_soa/include -isystem /Users/runner/work/algebra-plugins/algebra-plugins/build/_deps/googletest-src/googletest/include -isystem /Users/runner/work/algebra-plugins/algebra-plugins/build/_deps/googletest-src/googletest -isystem /Users/runner/work/algebra-plugins/algebra-plugins/build/_deps/vc-src -I/Users/runner/work/algebra-plugins/algebra-plugins/build/build/algebra_test_vc_soa.build/Release/DerivedSources-normal/x86_64 -I/Users/runner/work/algebra-plugins/algebra-plugins/build/build/algebra_test_vc_soa.build/Release/DerivedSources/x86_64 -I/Users/runner/work/algebra-plugins/algebra-plugins/build/build/algebra_test_vc_soa.build/Release/DerivedSources -F/Users/runner/work/algebra-plugins/algebra-plugins/build/bin/Release -Wall -Wextra -Wshadow -Wunused-local-typedefs -pedantic -Werror -DNDEBUG -std\=c++17 -Wno-local-type-template-args -Wno-unnamed-type-template-args -ffp-contract\=fast -march\=ivybridge -msse2 -msse3 -mssse3 -msse4.1 -msse4.2 -mavx -mno-sse4a -mno-fma -mno-bmi2 -mno-avx2 -mno-xop -mno-fma4 -mno-avx512f -mno-avx512vl -mno-avx512pf -mno-avx512er -mno-avx512cd -mno-avx512dq -mno-avx512bw -mno-avx512ifma -mno-avx512vbmi -MMD -MT dependencies -MF /Users/runner/work/algebra-plugins/algebra-plugins/build/build/algebra_test_vc_soa.build/Release/Objects-normal/x86_64/vc_soa.d --serialize-diagnostics /Users/runner/work/algebra-plugins/algebra-plugins/build/build/algebra_test_vc_soa.build/Release/Objects-normal/x86_64/vc_soa.dia -c /Users/runner/work/algebra-plugins/algebra-plugins/tests/vc_soa/vc_soa.cpp -o /Users/runner/work/algebra-plugins/algebra-plugins/build/build/algebra_test_vc_soa.build/Release/Objects-normal/x86_64/vc_soa.o
```

as my laptop does:

```
[zsh][tauriel]:algebra > more build/tests/CMakeFiles/algebra_test_vc_soa.dir/flags.make 
# CMAKE generated file: DO NOT EDIT!
# Generated by "Unix Makefiles" Generator, CMake Version 3.27

# compile CXX with /Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin/c++
CXX_DEFINES = 

CXX_INCLUDES = -I/Users/krasznaa/ATLAS/projects/algebra/algebra-plugins/storage/vc_soa/include -I/Users/krasznaa/ATLAS/projects/algebra/algebra-plugins/common/include -I/Users/krasznaa/ATLAS/projects/algebra/algebra-plugins/storage/common/include -I/Users/krasznaa/ATLAS/projects/algebra/algebra-plugins/frontend/vc_soa/include -I/Users/krasznaa/ATLAS/projects/algebra/algebra-plugins/math/cmath/include -I/Users/krasznaa/ATLAS/projects/algebra/algebra-plugins/math/common/include -I/Users/krasznaa/ATLAS/projects/algebra/algebra-plugins/math/vc_soa/include -isystem /Users/krasznaa/ATLAS/projects/algebra/build/_deps/googletest-src/googletest/include -isystem /Users/krasznaa/ATLAS/projects/algebra/build/_deps/googletest-src/googletest -isystem /Users/krasznaa/ATLAS/projects/algebra/build/_deps/vc-src

CXX_FLAGS =  -Wall -Wextra -Wshadow -Wunused-local-typedefs -pedantic -Werror -O3 -DNDEBUG -std=c++17 -isysroot /Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX12.1.sdk -mmacosx-version-min=11.7 -Wno-local-type-template-args -Wno-unnamed-type-template-args -ffp-contract=fast -march=haswell -msse2 -msse3 -mssse3 -msse4.1 -msse4.2 -mavx -mfma -mbmi2 -mavx2 -mno-sse4a -mno-xop -mno-fma4 -mno-avx512f -mno-avx512vl -mno-avx512pf -mno-avx512er -mno-avx512cd -mno-avx512dq -mno-avx512bw -mno-avx512ifma -mno-avx512vbmi
```

As a final thing I downloaded the binaries produced by the CI, from: https://github.com/krasznaa/algebra-plugins/actions/runs/6535189458 But unfortunately, even after I told macOS over and over again for every binary in that ZIP file that it's allowed to use it, I still get:

```
[zsh][tauriel]:Release > DYLD_LIBRARY_PATH=/Users/krasznaa/ATLAS/projects/algebra/ci/lib/Release ./algebra_test_vc_soa
dyld: Symbol not found: __ZNKSt3__115basic_stringbufIcNS_11char_traitsIcEENS_9allocatorIcEEE3strEv
  Referenced from: /Users/krasznaa/ATLAS/projects/algebra/ci/lib/Release/libgtest.1.14.0.dylib (which was built for Mac OS X 12.7)
  Expected in: /usr/lib/libc++.1.dylib

zsh: abort      DYLD_LIBRARY_PATH=/Users/krasznaa/ATLAS/projects/algebra/ci/lib/Release 
[zsh][tauriel]:Release > 
```

Since macOS is just not forward compatible. 😦

The flags used in both builds seem correct to me. 🤔 They both go up to AVX instructions. And according to Wikipedia, Icy Bridge is supposed to be able to do that. 😕 

So at this point I'll just do one more thing, I'll turn off the Vc compiler optimizations for a test build, and see if that works. If it does, we should just turn off those platform specific optimizations on macOS, and not the tests as a whole...

Reply:
Huhh... There is a twist. 😕 After I switched off the platform specific optimizations for this project's code with https://github.com/krasznaa/algebra-plugins/commit/0f7276e790c6468826b829eb5c54fce577883af8, the test still dies with an invalid instruction. 🤔

https://github.com/krasznaa/algebra-plugins/actions/runs/6535685632/job/17745684927

So it seems that it's something with Vc itself. 🤔 It seems that it's unable to choose the correct implementation of one of its optimized functions in that build. Interesting...

I now blame macOS 12.7 for not choosing the correct functions at runtime. One more thing that I'll try, after that I'll be out of ideas...

Reply:
Running the code through LLDB produces no error, so I strongly suspect that this is not actually related to compiler flags, but rather that we are producing some undefined behaviour somewhere. I'll try again with ASLR enabled to see if that re-introduces the bug.

Reply:
You mean you ran the debugger in the CI?

Because if not, then I'm very surprised that we're still not on the same page with this. 😕 The binary produced by the CI clearly contains an instruction that the build node's Ivy Bridge cpu cannot handle.

We've seen a similar issue in the offline a while ago. That's why "invalid instruction" seems pretty clear to me. I don't see signs of undefined behviour anywhere here...

Reply:
Yes, I'm running LLDB in the CI. :slightly_smiling_face:

Reply:
Okay, so here's the fun stuff. LLDB in the CI returns the following:

```
Running main() from /Users/runner/work/algebra-plugins/algebra-plugins/build/_deps/googletest-src/googletest/src/gtest_main.cc
[==========] Running 2 tests from 1 test suite.
[----------] Global test environment set-up.
[----------] 2 tests from test_vc_host
[ RUN      ] test_vc_host.vc_soa_vector
[       OK ] test_vc_host.vc_soa_vector (0 ms)
[ RUN      ] test_vc_host.vc_soa_getter
Process 4114 stopped
* thread #1, queue = 'com.apple.main-thread', stop reason = EXC_BAD_INSTRUCTION (code=EXC_I386_INVOP, subcode=0x0)
    frame #0: 0x00000001010da1b2 algebra_test_vc_soa`Vc_1::Vector<float, Vc_1::VectorAbi::Avx> Vc_1::Common::Trigonometric<Vc_1::ImplementationT<6u> >::atan2<Vc_1::Vector<float, Vc_1::VectorAbi::Avx> >(Vc_1::Vector<float, Vc_1::VectorAbi::Avx> const&, Vc_1::Vector<float, Vc_1::VectorAbi::Avx> const&) + 242
algebra_test_vc_soa`Vc_1::Common::Trigonometric<Vc_1::ImplementationT<6u> >::atan2<Vc_1::Vector<float, Vc_1::VectorAbi::Avx> >:
->  0x1010da1b2 <+242>: vfmsub231ps %ymm14, %ymm13, %ymm0     ; ymm0 = (ymm13 * ymm14) - ymm0
    0x1010da1b7 <+247>: vbroadcastss 0x38(%rax), %ymm14
    0x1010da1bd <+253>: vfmadd231ps %ymm0, %ymm13, %ymm14     ; ymm14 = (ymm13 * ymm0) + ymm14
    0x1010da1c2 <+258>: vbroadcastss 0x3c(%rax), %ymm0
Target 0: (algebra_test_vc_soa) stopped.
Process 4114 launched: '/Users/runner/work/algebra-plugins/algebra-plugins/build/bin/Release/algebra_test_vc_soa' (x86_64)
```

Interesting thing to note here:

1. The process is failing on `vfmsub231ps` which is a FMA3 instruction which Ivy Bridge does not support, so it is not a jump to an invalid instruction as I suspected.
2. The implementation identifier is 6, which [translates to AVX](https://github.com/VcDevel/Vc/blob/93d646eafb6714164eb28eb9571cbca6985572a0/Vc/global.h#L496), which is correct.
3. The `0x20000` bit [which indicates FMA3](https://github.com/VcDevel/Vc/blob/93d646eafb6714164eb28eb9571cbca6985572a0/Vc/global.h#L526) is _not_ set, which is correct.
4. Vc is still emitting a `vfmsub231ps` instruction in its implementation of `atan2` for AVX architectures without FMA.
5. It seems likely, but not certain, that [this subtraction](https://github.com/VcDevel/Vc/blob/1.4.3/src/trigonometric.cpp#L556) is the cause of the incorrect code emission.
6. The Vc code base does not appear to contain any mentions of `fmsub` intrinsics, so I would assume the code is automatically generated.

Thus:

1. Vc is jumping into the correct functions at runtime.
2. The assembly code for `libVc.a` is being generated incorrectly.

Probably more interesting things to be found but not today.

Reply:
Well, things turned out to be too much fun to leave until tomorrow. :smile: I haven't tested it, but I am quite certain I understand what is going on; it's a case of the compiler being smarter than the build system compiler.

First, a primer on how Vc is built. Vc produces a fat binary object `libVc.a` which is designed to run anywhere, and to run everywhere efficiently. Because the machine on which the code will be run is not known at compile time, `libVc.a` must contain multiple versions of all of its functions, for SSE2, SSE4.2, AVX, AVX2, etc.

Because the infrastructure for generating multiple versions of functions in the same translation unit is relatively poorly defined in most compilers (although some can do it), Vc achieves fat binary production by producing a large number of small binaries and linking them together. The trigonometry part of Vc is compiled into the following objects:

* `trigonometric_AVX+XOP+FMA.o`
* `trigonometric_AVX+XOP+FMA4.o`
* `trigonometric_SSE4_1.o`
* `trigonometric_SSE3.o`
* `trigonometric_SSSE3.o`
* `trigonometric_SSE+XOP+FMA.o`
* `trigonometric_AVX+FMA.o`
* `trigonometric_AVX2+FMA+BMI2.o`
* `trigonometric_AVX.o`
* `trigonometric_SSE+XOP+FMA4.o`
* `trigonometric_SSE2.o`

All of this is well and good. Let's inspect the `atan2` function (which we know to be problematic) in the `trigonometric_AVX.o` file, which is the one that will be used on the Github runners:

```
00000000000020f0 <Vc_1::Vector<float, Vc_1::VectorAbi::Avx> Vc_1::Common::Trigonometric<Vc_1::ImplementationT<6u>>::atan2<Vc_1::Vector<float, Vc_1::VectorAbi::Avx>>(Vc_1::Vector<float, Vc_1::VectorAbi::Avx> const
&, Vc_1::Vector<float, Vc_1::VectorAbi::Avx> const&)>:
[some assembly redacted for brevity]
    21d6: c4 62 7d 18 70 30             vbroadcastss    48(%rax), %ymm14
    21dc: c4 41 14 59 f6                vmulps  %ymm14, %ymm13, %ymm14
    21e1: c4 e2 7d 18 60 34             vbroadcastss    52(%rax), %ymm4
    21e7: c5 8c 5c e4                   vsubps  %ymm4, %ymm14, %ymm4
    21eb: c5 94 59 e4                   vmulps  %ymm4, %ymm13, %ymm4
    21ef: c4 62 7d 18 70 38             vbroadcastss    56(%rax), %ymm14
    21f5: c5 8c 58 e4                   vaddps  %ymm4, %ymm14, %ymm4
    21f9: c5 94 59 e4                   vmulps  %ymm4, %ymm13, %ymm4
    21fd: c4 62 7d 18 70 3c             vbroadcastss    60(%rax), %ymm14
    2203: c4 c1 5c 5c e6                vsubps  %ymm14, %ymm4, %ymm4
    2208: c5 94 59 e4                   vmulps  %ymm4, %ymm13, %ymm4
```

Looks good! This is all AVX code which does not use FMA. All is well, and this would run perfectly on the CI runner. Note the instructions that are being run: vector subtractions (`vsubps`) and additions (`vaddps`) followed by multiplications (`vmulps`). Let's find the same code in the fat binary `libVc.a`:

```
0000000000002460 <Vc_1::Vector<float, Vc_1::VectorAbi::Avx> Vc_1::Common::Trigonometric<Vc_1::ImplementationT<7u>>::atan2<Vc_1::Vector<float, Vc_1::VectorAbi::Avx>>(Vc_1::Vector<float, Vc_1::VectorAbi::Avx> const
&, Vc_1::Vector<float, Vc_1::VectorAbi::Avx> const&)>:
[some assembly redacted for brevity]
    2547: c4 62 7d 18 70 30             vbroadcastss    48(%rax), %ymm14
    254d: c4 e2 7d 18 40 34             vbroadcastss    52(%rax), %ymm0
    2553: c4 c2 15 ba c6                vfmsub231ps     %ymm14, %ymm13, %ymm0 ## ymm0 = (ymm13 * ymm14) - ymm0
    2558: c4 62 7d 18 70 38             vbroadcastss    56(%rax), %ymm14
    255e: c4 62 15 b8 f0                vfmadd231ps     %ymm0, %ymm13, %ymm14 ## ymm14 = (ymm13 * ymm0) + ymm14
    2563: c4 e2 7d 18 40 3c             vbroadcastss    60(%rax), %ymm0
    2569: c4 c2 15 ba c6                vfmsub231ps     %ymm14, %ymm13, %ymm0 ## ymm0 = (ymm13 * ymm14) - ymm0
    256e: c5 94 59 c0                   vmulps  %ymm0, %ymm13, %ymm0
```

Two things to notice here. First, the code has _changed_ during linking. Second, the code is _equivalent_ to the original code. How did this happen?

The answer is link-time optimization. Vc takes great care to ensure that the individual translation units are compiled with the right settings, but the linker is _not_ told to respect any of this; the linker is free to optimize at it pleases. In this case, the linker has performed link-time peephole optimization or, more specifically, link-time floating point contraction, which is the contraction of additions (or subtractions) and multiplications into fused multiply-addition operations!

How cool is that? :smile: Presumably the Mach-O binary format used by XCode does not store information about compilation targets to ensure that this doesn't happen, but I don't know.

Reply:
Confirmed, the fix was extremely simple:

First, I had to patch Vc to disable link-time floating point contraction when using XCode, which I did in https://github.com/stephenswat/Vc/commit/d56339236d9014c6e78e673df4450c977b04a078

Second, I updated the version of Vc used to my patched version in https://github.com/stephenswat/algebra-plugins/commit/e0c40e168588ec5f41740e9af367338802cfdda3

This happily passes the CI: https://github.com/stephenswat/algebra-plugins/actions/runs/6540226183/job/17759772680

Reply:
Nvm, I just found it :D

Reply:
@stephenswat dude.... 😮

Reply:
As it seems that this is a bit stuck with VC, should we add the flag here for corresponding version of xcode?

Reply:
Since none of us were active on providing a better fix for this, let's just keep the benchmark turned off on macOS for now. We can fix this in a better way in the future, based on what we figured out last year...